nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ParameterContextResource.java

```diff
@@ -0,0 +1,1185 @@
+            throw new IllegalArgumentException("Parameter Context's ID must be specified");
+        }
+        if (requestEntity.getId() == null) {
+            throw new IllegalArgumentException("Entity's ID must be specified");
+        }
+
+        // We will perform the updating of the Parameter Context in a background thread because it can be a long-running process.
+        // In order to do this, we will need some objects that are only available as Thread-Local variables to the current
+        // thread, so we will gather the values for these objects up front.
+        final boolean replicateRequest = isReplicateRequest();
+        final ComponentLifecycle componentLifecycle = replicateRequest ? clusterComponentLifecycle : localComponentLifecycle;
+        final NiFiUser user = NiFiUserUtils.getNiFiUser();
+
+        // Workflow for this process:
+        // 1. Determine which components will be affected and are enabled/running
+        // 2. Verify READ and WRITE permissions for user, for every component that is affected
+        // 3. Verify READ and WRITE permissions for user, for Parameter Context
+        // 4. Stop all Processors that are affected.
+        // 5. Wait for all of the Processors to finish stopping.
+        // 6. Disable all Controller Services that are affected.
+        // 7. Wait for all Controller Services to finish disabling.
+        // 8. Update Parameter Context
+        // 9. Re-Enable all affected Controller Services
+        // 10. Re-Start all Processors
+
+        final Set<AffectedComponentEntity> affectedComponents = serviceFacade.getComponentsAffectedByParameterContextUpdate(contextDto);
+        logger.debug("Received Update Request for Parameter Context: {}; the following {} components will be affected: {}", requestEntity, affectedComponents.size(), affectedComponents);
+
+        final InitiateChangeParameterContextRequestWrapper requestWrapper = new InitiateChangeParameterContextRequestWrapper(requestEntity, componentLifecycle, getAbsolutePath(),
+            affectedComponents, replicateRequest, user);
+
+        final Revision requestRevision = getRevision(requestEntity.getRevision(), contextDto.getId());
+        return withWriteLock(
+            serviceFacade,
+            requestWrapper,
+            requestRevision,
+            lookup -> {
+                // Verify READ and WRITE permissions for user, for the Parameter Context itself
+                final Authorizable parameterContext = lookup.getParameterContext(requestEntity.getId());
+                parameterContext.authorize(authorizer, RequestAction.READ, user);
+                parameterContext.authorize(authorizer, RequestAction.WRITE, user);
+
+                // Verify READ and WRITE permissions for user, for every component that is affected
+                for (final AffectedComponentEntity entity : affectedComponents) {
+                    final AffectedComponentDTO dto = entity.getComponent();
```

Comment:
Good catch.

Reply:
Why isn't the new approach like the `getActiveComponentsAffectedByVariableRegistryUpdate` where we directly return the DTO? This way we wouldn't need to deal with the conditionally nulled out component. I think that would be fine in this case given its usage.

Reply:
Actually, now that I am more closing testing and evaluating I believe my previous suggestion of only needing to authorize against active components was wrong. We take this approach with variables because the variables are referenced by name. Because of this, an EL statement could refer to a variable or a flowfile attribute. What actually resolves will not be known until the flow is actually executed. Because of this, we do not know definitively if EL references a variable or not.

However, with parameters, we have eliminated this ambiguity and we can now authorize against this too. This is what is allowing us to support sensitive parameters. Because of this, we must authorize against all affected components and not just the active ones. Sorry for the misdirection.